main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala

```diff
@@ -0,0 +1,198 @@
+                    } else {
+                      expected match {
+                        // If the expected effect is Pure, we check if the concrete effect is Pure.
+                        case Type.Pure => {
+                          if (concreteEff == Symbol.mkEffectSym("Pure")) {
+                            println("Found expected pure effect")
+                            val newPath = new EffectProvenancePath.Path(concreteEff, path.reverse.head, constr :: path, constr)
+                            printPath(newPath)
+                          }
+                        }
+                        case _ => ()
+                      }
+                    }
+                  }
+                }
+              }
+            }
+            case _ => ()
+          }
+        }
+        case _ => ()
+      }
+    }
+  }
+
+  // Find concrete effects to be introduced and those to be removed
+  private def findConcreteEffects(tpe: Type): (SortedSet[Symbol.EffSym], SortedSet[Symbol.EffSym]) = {
+    tpe match {
+      case Type.Apply(Type.Apply(Type.Difference, t1, _), t2, _) => {
+        val (add1, sub1) = findConcreteEffects(t1)
+        val (add2, sub2) = findConcreteEffects(t2)
+        (add1.union(sub2), sub1.union(add2))
+      }
+      case Type.Apply(Type.Apply(Type.Union, t1, _), t2, _) => {
+        val (add1, sub1) = findConcreteEffects(t1)
+        val (add2, sub2) = findConcreteEffects(t2)
+        (add1.union(add2), sub1.union(sub2))
+      }
+      case Type.Cst(TypeConstructor.Effect(sym, _), _) => (SortedSet(sym), SortedSet.empty)
+      case _ => (SortedSet.empty, SortedSet.empty)
+    }
+  }
+
+  // Prints the full path of constraints indicating which variables
+  // were traversed and the constraints at each step.
+  private def printPath(path: EffectProvenancePath.Path): Unit = {
+    path match {
+      case EffectProvenancePath.Path(sym, source, elms, sink) => {
+        var msg: String = "///////////////////////////////////////////////////////////////////////////////\n"
+        msg += s"Found path!!!\n"
+        msg += "///////////////////////////////////////////////////////////////////////////////\n"
+        msg += s"Source: ${source}\n"
+        msg += "Constraints:\n"
+        elms.reverse.foreach { elm =>
+          msg += s"  -> ${elm} at ${elm.loc.format}\n"
+        }
+        msg += s"Sink: ${sink}\n"
+        msg += "///////////////////////////////////////////////////////////////////////////////\n"
+        println(msg)
+      }
+    }
+  }
+
+  // Dump all the constraints with their provenance.
+  private def printConstrs(constrs: List[TypeConstraint]): Unit = {
+    constrs.foreach {
+      case TypeConstraint.Equality(eff1, eff2, prov) => prov match {
+        case _ => println(s"$eff1 ~ $eff2 at ${prov.loc} with provenance $prov")
+      }
+      case _ => ()
+    }
+  }
+
+  // Represents a path through the constraints from a source to a sink.
+  object EffectProvenancePath {
```

Comment:
Will there be more than one kind of path in the future? If not, we can just turn this into a case class directly.